File: detect_english/python_code/detector.py
from os import getcwd
from os.path import exists, join


class EnglishDetector(object):
    """
    A class for detecting English-like text.

    Attributes:
        text (str): The text to be analyzed.
        _english_words (dictionary): A set containing English words for quick lookup.

    Methods:
        __init__(): Initializes the class attributes.
        load_dictionary(dictionary_path="dictionary.txt"): Loads a dictionary of English words from a file.
        set_text(text): Sets the text to be analyzed.
        is_valid_english(word_percentage_match=20, letters_percentage_match=80):
            Determines if the text resembles English based on the percentage of English words and letters.
        _get_num_of_english_word_percentage(): Calculates the percentage of English words in the text.
    """

    def __init__(self) -> None:
        self.text           = ""
        self._english_words = {}
        self.file_path      = ""

    def load_dictionary(self):
        """
        Loads a dictionary of English words from a file.

        Args:
            dictionary_path (str, optional): The path to the dictionary file. Default is "dictionary.txt".
        """
        running = True

        try:
            with open(self.file_path) as f:
                while running:
                    word = f.readline().strip()
                    if not word:
                        running = False
                    else:
                        self._english_words[word.lower()] = True
        except Exception as e:
            raise Exception(f"Expected dictionary but got {e} error")

    def set_text(self, text):
        """
        Sets the text to be analyzed.

        Args:
            text (str): The text to be analyzed.
        """
        self.text = text

    def set_dictionary_file_path(self, file_path):
        """
        Sets the file path
        Args:
            text (str): The file path
        """
        if not exists(file_path):
            raise Exception("The file was not found!!")
        self.file_path = file_path

    def is_valid_english(self, word_percentage_match=75):
        """
        Determines if the text resembles English based on the percentage of English words and letters.

        Args:
            word_percentage_match (int, optional): The minimum percentage of English words required for a match.
                Default is 75.

        Returns:
            bool: True if the text resembles English, False otherwise.
        """
        return self._get_num_of_english_word_percentage() >= word_percentage_match

    def _get_num_of_english_word_percentage(self):
        """
        Calculates the percentage of English words in the text.

        Returns:
            float: The percentage of English words in the text.
        """
        words = self.text.split()
        percentage = 100

        if not words:
            return 0
        word_match = sum(
            (
                1
                for word in words
                if self._english_words.get(word.lower().strip(), False)
            )
        )

        return (word_match / len(words)) * percentage


if __name__ == "__main__":

    text = "This is the way of the dragon egbie"
    dictionary_file_path = join(getcwd(), "detect_english", "dictionary.txt")

    detector = EnglishDetector()
    detector.set_text(text)
    detector.set_dictionary_file_path(dictionary_file_path)
    detector.load_dictionary()

    is_valid_english = detector.is_valid_english()
    print(f"is the passage a valid English text: {is_valid_english}")

    text = "???????????????????????????????dd??? word word"
    detector.set_text(text)
    is_valid_english = detector.is_valid_english()
    print(f"is the passage a valid English text: {is_valid_english}")

File: detect_english/js/detector.mjs
import path from "path";
import fs from "fs";

/**
 * A class to detect the presence of English words in a given text.
 */
class EnglishDetector {
   
    constructor() {
       
        this._englishWords = null;
        this._filePath     = null;
        this.text          = null;
    }

    /**
     * Sets the file path of the dictionary.
     * @param {string} filePath - The path to the dictionary file.
     * @returns {Promise<void>}
     */
    async setDictionaryFilePath(filePath) {
        await this._doesFileExist(filePath);
        this._filePath = filePath;
    }

    /**
     * Sets the text to be analyzed.
     * @param {string} text - The text to be analyzed.
     */
    setText(text) {
        this.text = text;
    }

    /**
     * Checks if the provided text contains a valid percentage of English words.
     * @param {number} [percentageWordMatch=75] - The required percentage of English words for validity.
     * @returns {boolean} - True if the text contains a valid percentage of English words, otherwise false.
     */
    isValidEnglish(percentageWordMatch = 75) {
        return this._getNumOfEnglishWordPercentage() >= percentageWordMatch;
    }

    /**
     * Calculates the percentage of English words in the provided text.
     * @private
     * @returns {number} - The percentage of English words.
     */
    _getNumOfEnglishWordPercentage() {
        let match = 0;
        const words = this.text.split(" ");
        const percentage = 100;

        for (let word of words) {
            if (this._englishWords[word]) {
                match++;
            }
        }
        return (match / words.length) * percentage;
    }

    /**
     * Loads the dictionary from the specified file path.
     * @returns {Promise<void>}
     */
    async loadDictionary() {
        try {
            const fileContent = await fs.promises.readFile(this._filePath, 'utf8');
            this.populateEnglishWordsFromContent(fileContent);
        } catch (error) {
            console.error(`Error expected file but received an error of ${error}`);
        }
    }

    /**
     * Populates the English words dictionary from the provided file content.
     * This method parses the content of the dictionary file and adds each word to the English words dictionary.
     * Each word is converted to lowercase and trimmed before being added to the dictionary.
     * @param {string} fileContent - The content of the dictionary file, where each line represents a word.
     *                             
     * @private
     */
    populateEnglishWordsFromContent(fileContent) {
        if (fileContent) {
            for (let word of fileContent.split("\n")) {
                word = word.toLowerCase().trim();
                if (word) {
                    this._englishWords[word] = word;
                }
            }
        }
    }

    /**
     * Checks if the specified file exists.
     * @param {string} filePath - The path to the file.
     * @returns {Promise<void>}
     * @private
     */
    async _doesFileExist(filePath) {
        try {
            await fs.promises.access(filePath, fs.constants.F_OK);
        } catch (error) {
            console.error("The file could not be found!!");
            throw error;
        }
    }
}

export default EnglishDetector;

// async function main() {
//     const DICTIONARY_FILE_PATH = path.join(process.cwd(), "detect_english", "dictionary.txt");
//     const text  = "the way of the dragon? word word";

//     const englishDetector = new EnglishDetector();
//     englishDetector.setText(text);

//     await englishDetector.setDictionaryFilePath(DICTIONARY_FILE_PATH);
//     await englishDetector.loadDictionary();

//     const percentage = englishDetector.isValidEnglish();
//     console.log(percentage)
    
// }

// main().catch(error => {
//     console.error('An error occurred:', error);
// });
